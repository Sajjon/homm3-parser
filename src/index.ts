import { parse, StreamInput, TagOrWrapper } from 'binary-markup'
import { inflate } from 'pako'
import { h3mFile, H3MFile } from './parsers/h3m/h3m.bml'

export * from './parsers/assets'

export { H3MFile } from './parsers/h3m/h3m.bml'

export function parseH3M(data: StreamInput): H3MFile {
  return parse<H3MFile>(h3mFile, data)
}

export function parseH3MFile(data: Uint8Array): H3MFile {
  const content = inflate(data)
  return parseH3M(content)
}

export function parseAsset<T>(type: TagOrWrapper<T>, data: StreamInput): T {
  return parse(type, data)
}
