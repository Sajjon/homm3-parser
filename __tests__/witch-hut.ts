import { testMap } from './utils'
import { H3MFile } from '../src'
import { WitchHut } from '../src/parsers/h3m/objectDetails/witchHut.bml'

describe('witch hut', () => {
  testMap('witch hut', 'witch-hut.h3m', (map: H3MFile) =>
    map.objectDetails.entries.map(obj => {
      const witchHut = obj.body as WitchHut
      return {
        x: obj.x,
        y: obj.y,
        skills: witchHut.potentialSkills,
      }
    }),
  )
})
