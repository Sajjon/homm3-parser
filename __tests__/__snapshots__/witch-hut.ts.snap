// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`witch hut should parse witch hut 1`] = `
Array [
  Object {
    "skills": Array [
      "AirMagic",
    ],
    "x": 2,
    "y": 2,
  },
  Object {
    "skills": Array [
      "Archery",
    ],
    "x": 5,
    "y": 2,
  },
  Object {
    "skills": Array [
      "Armorer",
    ],
    "x": 8,
    "y": 2,
  },
  Object {
    "skills": Array [
      "Artillery",
    ],
    "x": 11,
    "y": 2,
  },
  Object {
    "skills": Array [
      "Ballistics",
    ],
    "x": 14,
    "y": 2,
  },
  Object {
    "skills": Array [
      "Diplomacy",
    ],
    "x": 17,
    "y": 2,
  },
  Object {
    "skills": Array [
      "EagleEye",
    ],
    "x": 20,
    "y": 2,
  },
  Object {
    "skills": Array [
      "EarthMagic",
    ],
    "x": 23,
    "y": 2,
  },
  Object {
    "skills": Array [
      "Estates",
    ],
    "x": 26,
    "y": 2,
  },
  Object {
    "skills": Array [
      "FireMagic",
    ],
    "x": 29,
    "y": 2,
  },
  Object {
    "skills": Array [
      "FirstAid",
    ],
    "x": 32,
    "y": 2,
  },
  Object {
    "skills": Array [
      "Intelligence",
    ],
    "x": 35,
    "y": 2,
  },
  Object {
    "skills": Array [
      "Leadership",
    ],
    "x": 2,
    "y": 5,
  },
  Object {
    "skills": Array [
      "Learning",
    ],
    "x": 5,
    "y": 5,
  },
  Object {
    "skills": Array [
      "Logistics",
    ],
    "x": 8,
    "y": 5,
  },
  Object {
    "skills": Array [
      "Luck",
    ],
    "x": 11,
    "y": 5,
  },
  Object {
    "skills": Array [
      "Mysticism",
    ],
    "x": 14,
    "y": 5,
  },
  Object {
    "skills": Array [
      "Navigation",
    ],
    "x": 17,
    "y": 5,
  },
  Object {
    "skills": Array [
      "Necromancy",
    ],
    "x": 20,
    "y": 5,
  },
  Object {
    "skills": Array [
      "Offense",
    ],
    "x": 23,
    "y": 5,
  },
  Object {
    "skills": Array [
      "Pathfinding",
    ],
    "x": 26,
    "y": 5,
  },
  Object {
    "skills": Array [
      "Resistance",
    ],
    "x": 29,
    "y": 5,
  },
  Object {
    "skills": Array [
      "Scholar",
    ],
    "x": 32,
    "y": 5,
  },
  Object {
    "skills": Array [
      "Scouting",
    ],
    "x": 35,
    "y": 5,
  },
  Object {
    "skills": Array [
      "Sorcery",
    ],
    "x": 2,
    "y": 8,
  },
  Object {
    "skills": Array [
      "Tactics",
    ],
    "x": 5,
    "y": 8,
  },
  Object {
    "skills": Array [
      "AirMagic",
      "WaterMagic",
    ],
    "x": 8,
    "y": 8,
  },
  Object {
    "skills": Array [
      "Wisdom",
      "AirMagic",
    ],
    "x": 11,
    "y": 8,
  },
]
`;
